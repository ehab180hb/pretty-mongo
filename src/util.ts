/**
 * Will check whether or not the given object is empty
 */
export const isEmptyObject = (obj: object) => {
  return obj && obj.constructor === Object && Object.keys(obj).length === 0
}

/**
 * Will remove the fields that do not meet the given condition
 */
export const removeValues = (o: object, condition: (val: any) => boolean) => {
  const newObj = {}
  for (const k in o) if (!condition(o[k])) newObj[k] = o[k]
  return newObj
}

/**
 * Returns an array with ranges from arg 1 up to (but not including) arg 2
 * @example
 * getRange(3, 6) // [ 3, 4, 5 ]
 */
export const getRange = (from: number, to: number): number[] => {
  if (from > to) throw new Error(`Value of 'from' should be lower than or equal to 'to'`)
  return Array.from({ length: to - from }, (_v, k) => k + from)
}
