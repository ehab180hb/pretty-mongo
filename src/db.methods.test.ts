import * as dbTestUtil from 'mongo-test-suit'
import { ObjectId } from 'bson'
import { getRange } from './util'
import { MongoCollectionMethods } from './db.methods'
import * as randomString from 'randomstring'

const dbName = `my-db-${randomString.generate(7)}`
const uri = `mongodb://localhost/${dbName}`

const { fillCollection, getCollectionSnapshot, getTestDb, initializeTestDB, flushDb } = dbTestUtil

const collectionName = 'testCollection'
let myService: MongoCollectionMethods<any>

beforeAll(async () => {
  await initializeTestDB(uri)
  myService = new MongoCollectionMethods(getTestDb().collection(collectionName))
})

beforeEach(async () => {
  await flushDb()
})

const replaceIdWithTs = (docs: { _id: ObjectId }[]) => {
  return docs.map(x => {
    return {
      ...x,
      _id: x._id.getTimestamp(),
    }
  })
}

describe('DB methods tests', () => {
  describe('Storing data', () => {
    describe('createOne()', () => {
      test('Creates record properly', async () => {
        const r = await myService.createOne({ hi: 'hello' })
        expect(r).toHaveProperty('toHexString')
        expect(await getCollectionSnapshot({ collectionName, removeIds: true })).toMatchSnapshot()
      })
    })

    describe('updateOne()', () => {
      const before = { email: 'IExist@gmail.com' }
      const after = { email: 'IExist2@gmail.com' }
      test('When record exists', async () => {
        await fillCollection(collectionName, [before, { email: 'meToo@gmail.com' }])
        const r = await myService.updateOne(before, after)
        expect(r).toBeTruthy()
        expect(await getCollectionSnapshot({ collectionName, removeIds: true })).toMatchSnapshot()
      })
      test('When an object in the update is undefined', async () => {
        await fillCollection(collectionName, [
          { ...before, wontBeDeleted: 'x', willBeDeleted: 'y' },
          { email: 'meToo@gmail.com' },
        ])
        const r = await myService.updateOne(before, { ...after, willBeDeleted: undefined })
        expect(r).toBeTruthy()
        expect(await getCollectionSnapshot({ collectionName, removeIds: true })).toMatchSnapshot()
      })
      test('When record does not exist', async () => {
        const r = await myService.updateOne(before, after)
        expect(r).toBeFalsy()
        expect(await getCollectionSnapshot({ collectionName })).toMatchSnapshot()
      })
    })

    describe('updateFieldsWithSpecialOp()', () => {
      test('Without options', async () => {
        await fillCollection(collectionName, [{ x: 'y', num: 1 }])
        await myService.updateFieldsWithSpecialOp({ x: 'y' }, { $inc: { num: 1 } })
        expect(await getCollectionSnapshot({ collectionName, removeIds: true })).toMatchSnapshot()
      })
    })

    describe('createMany()', () => {
      test('When given the _id', async () => {
        const docs = getRange(0, 5).map(() => ({ _id: new ObjectId(), k: 'f' }))
        await myService.createMany(docs)
        const snap = (await getCollectionSnapshot({ collectionName })) as { _id: ObjectId }[]
        expect(replaceIdWithTs(snap)).toMatchSnapshot()
      })
      test('When not given the _id', async () => {
        const docs = getRange(0, 5).map(() => ({ v: 'n' }))
        await myService.createMany(docs)
        const snap = (await getCollectionSnapshot({ collectionName })) as { _id: ObjectId }[]
        expect(replaceIdWithTs(snap)).toMatchSnapshot()
      })
    })
  })

  describe('Fetching data()', () => {
    const data = [
      { _id: new ObjectId('5cc1db02dc067f5df93c5432'), hi: 'hello', same: 'same', name: 'Mark' },
      { _id: new ObjectId('5cc1db02dc067f5df93c5433'), tak: 'nie', same: 'same', name: 'John' },
    ]
    beforeEach(async () => {
      await fillCollection(collectionName, data)
    })
    describe('getOneById()', () => {
      test('When a record exists', async () => {
        const r = await myService.getOneById(data[0]._id)
        expect(r).toMatchSnapshot()
      })
      test('When a record exists with projection', async () => {
        const r = await myService.getOneById(data[0]._id, { projection: { name: 1 } })
        expect(r).toMatchSnapshot()
      })
      test('When a record does not exist', async () => {
        const r = await myService.getOneById(new ObjectId('5cc1db02dc067f5df93c5434'))
        expect(r).toBeUndefined()
      })
    })

    describe('getOne', () => {
      test('When it exists', async () => {
        expect(await myService.getOne({ name: 'Mark' })).toMatchSnapshot()
      })
      test('When it exists - with opts', async () => {
        expect(await myService.getOne({ name: 'Mark' }, { projection: { same: 0 } })).toMatchSnapshot()
      })
      test('When it does not exist', async () => {
        expect(await myService.getOne({ name: 'Sandra' })).toBeUndefined()
      })
    })
    describe('getMany()', () => {
      test('When fetching all', async () => {
        const r = await myService.getMany({})
        expect(r).toMatchSnapshot()
      })
      test('with projection', async () => {
        const r = await myService.getMany({ hi: 'hello' }, { projection: { same: 0 } })
        expect(r).toMatchSnapshot()
      })
      test('with sorting', async () => {
        const r = await myService.getMany({}, { sort: { name: 1 } })
        expect(r).toMatchSnapshot()
      })
      test('with skipping', async () => {
        const r = await myService.getMany({ same: 'same' }, { skip: 1 })
        expect(r).toMatchSnapshot()
      })
      test('When nothing matches', async () => {
        const r = await myService.getMany({ non: 'existing' })
        expect(r).toEqual([])
      })
    })
  })
  describe('getTotalCount()', () => {
    beforeEach(async () => {
      await fillCollection(collectionName, [{ x: 'y' }, { x: 'y' }, { x: 'y' }, { v: 'v' }])
    })
    test('When counting all', async () => {
      const r = await myService.getTotalCount()
      expect(r).toBe(4)
    })
    test('When counting some', async () => {
      const r = await myService.getTotalCount({ v: 'v' })
      expect(r).toBe(1)
    })
  })

  describe('Deleting data', () => {
    const data = [
      { _id: new ObjectId('5cc1db02dc067f5df93c5432'), hi: 'hello', same: 'same', name: 'Mark' },
      { _id: new ObjectId('5cc1db02dc067f5df93c5433'), tak: 'nie', same: 'same', name: 'John' },
    ]
    beforeEach(async () => {
      await fillCollection(collectionName, data)
    })
    describe('DeleteOne', () => {
      test('When record exists', async () => {
        const r = await myService.deleteOneById(data[0]._id)
        expect(r).toBe(1)
        expect(await getCollectionSnapshot({ collectionName })).toMatchSnapshot()
      })
      test('When record does not exist', async () => {
        const r = await myService.deleteOneById(new ObjectId('5cc1db02dc067f5df93c5620'))
        expect(r).toBeFalsy()
        expect(await getCollectionSnapshot({ collectionName })).toMatchSnapshot()
      })
    })
  })
})
