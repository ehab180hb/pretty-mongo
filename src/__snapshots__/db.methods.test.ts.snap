// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DB methods tests Deleting data DeleteOne When record does not exist 1`] = `
Array [
  Object {
    "_id": "5cc1db02dc067f5df93c5432",
    "hi": "hello",
    "name": "Mark",
    "same": "same",
  },
  Object {
    "_id": "5cc1db02dc067f5df93c5433",
    "name": "John",
    "same": "same",
    "tak": "nie",
  },
]
`;

exports[`DB methods tests Deleting data DeleteOne When record exists 1`] = `
Array [
  Object {
    "_id": "5cc1db02dc067f5df93c5433",
    "name": "John",
    "same": "same",
    "tak": "nie",
  },
]
`;

exports[`DB methods tests Fetching data() getMany() When fetching all 1`] = `
Array [
  Object {
    "_id": "5cc1db02dc067f5df93c5432",
    "hi": "hello",
    "name": "Mark",
    "same": "same",
  },
  Object {
    "_id": "5cc1db02dc067f5df93c5433",
    "name": "John",
    "same": "same",
    "tak": "nie",
  },
]
`;

exports[`DB methods tests Fetching data() getMany() with projection 1`] = `
Array [
  Object {
    "_id": "5cc1db02dc067f5df93c5432",
    "hi": "hello",
    "name": "Mark",
  },
]
`;

exports[`DB methods tests Fetching data() getMany() with skipping 1`] = `
Array [
  Object {
    "_id": "5cc1db02dc067f5df93c5433",
    "name": "John",
    "same": "same",
    "tak": "nie",
  },
]
`;

exports[`DB methods tests Fetching data() getMany() with sorting 1`] = `
Array [
  Object {
    "_id": "5cc1db02dc067f5df93c5433",
    "name": "John",
    "same": "same",
    "tak": "nie",
  },
  Object {
    "_id": "5cc1db02dc067f5df93c5432",
    "hi": "hello",
    "name": "Mark",
    "same": "same",
  },
]
`;

exports[`DB methods tests Fetching data() getOne When it exists - with opts 1`] = `
Object {
  "_id": "5cc1db02dc067f5df93c5432",
  "hi": "hello",
  "name": "Mark",
}
`;

exports[`DB methods tests Fetching data() getOne When it exists 1`] = `
Object {
  "_id": "5cc1db02dc067f5df93c5432",
  "hi": "hello",
  "name": "Mark",
  "same": "same",
}
`;

exports[`DB methods tests Fetching data() getOneById() When a record exists 1`] = `
Object {
  "_id": "5cc1db02dc067f5df93c5432",
  "hi": "hello",
  "name": "Mark",
  "same": "same",
}
`;

exports[`DB methods tests Fetching data() getOneById() When a record exists with projection 1`] = `
Object {
  "_id": "5cc1db02dc067f5df93c5432",
  "name": "Mark",
}
`;

exports[`DB methods tests Storing data createMany() When given the _id 1`] = `
Array [
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "k": "f",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "k": "f",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "k": "f",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "k": "f",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "k": "f",
  },
]
`;

exports[`DB methods tests Storing data createMany() When not given the _id 1`] = `
Array [
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "v": "n",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "v": "n",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "v": "n",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "v": "n",
  },
  Object {
    "_id": 2018-01-25T17:30:44.000Z,
    "v": "n",
  },
]
`;

exports[`DB methods tests Storing data createOne() Creates record properly 1`] = `
Array [
  Object {
    "hi": "hello",
  },
]
`;

exports[`DB methods tests Storing data updateFieldsWithSpecialOp() Without options 1`] = `
Array [
  Object {
    "num": 2,
    "x": "y",
  },
]
`;

exports[`DB methods tests Storing data updateOne() When an object in the update is undefined 1`] = `
Array [
  Object {
    "email": "IExist2@gmail.com",
    "wontBeDeleted": "x",
  },
  Object {
    "email": "meToo@gmail.com",
  },
]
`;

exports[`DB methods tests Storing data updateOne() When record does not exist 1`] = `Array []`;

exports[`DB methods tests Storing data updateOne() When record exists 1`] = `
Array [
  Object {
    "email": "IExist2@gmail.com",
  },
  Object {
    "email": "meToo@gmail.com",
  },
]
`;
